rust-runtime/aws-smithy-http-server/src/exception.rs

```diff
@@ -0,0 +1,104 @@
+//!
+//! Exception types' behavior doesn't have anything to do with how _programming language exceptions_
+//! behave: there are no exceptions in Rust, `[SmithyFrameworkException`] is just a regular Rust
+//! enum. Their name comes from their representation in all AWS protocols: for example,
+//! [`SmithyFrameworkExceptionType::Serialization`] renders an `X-Amzn-Errortype` header with the
+//! value `SerializationException` when using the RestJson1 protocol.
+//!
+//! The way generated code works is it always works with [`crate::rejection`] types when
+//! deserializing requests and serializing response. Just before a response needs to be sent, the
+//! generated code looks up and converts into the corresponding exception type, and then it uses
+//! the exception's [`axum_core::response::IntoResponse`] implementation to render and send a
+//! response.
+
+use crate::protocols::Protocol;
+
+#[derive(Debug)]
+pub enum SmithyFrameworkExceptionType {
+    // UnknownOperation,
+    Serialization(crate::Error),
+    // UnsupportedMediaType,
+    // NotAcceptable,
+}
+
+/// String representation of the exception type.
+/// Used as the value of the `X-Amzn-Errortype` header in RestJson1.
+/// Used as the value passed to construct an [`crate::ExtensionRejection`].
+impl SmithyFrameworkExceptionType {
+    pub fn name(&self) -> &'static str {
+        match self {
+            SmithyFrameworkExceptionType::Serialization(_) => "SerializationException",
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct SmithyFrameworkException {
+    pub protocol: Protocol,
+    pub exception_type: SmithyFrameworkExceptionType,
+}
+
+impl axum_core::response::IntoResponse for SmithyFrameworkException {
+    fn into_response(self) -> axum_core::response::Response {
+        let status_code = match self.exception_type {
+            SmithyFrameworkExceptionType::Serialization(_) => http::StatusCode::BAD_REQUEST,
+        };
+
+        let body = crate::body::to_boxed(match self.protocol {
+            Protocol::RestJson1 => "{}",
+            Protocol::RestXml => "",
+        });
+
+        let mut builder = http::Response::builder();
+        builder = builder.status(status_code);
+
+        match self.protocol {
+            Protocol::RestJson1 => {
+                builder = builder
+                    .header("Content-Type", "application/json")
+                    .header("X-Amzn-Errortype", self.exception_type.name());
+            }
+            Protocol::RestXml => {
+                builder = builder.header("Content-Type", "application/xml");
+            }
+        }
+
+        // TODO What extension type should we use here?
+        // TODO `ResponseExtensions` should probably be renamed, as it's something
+        // operation-specific. `SmithyFrameworkException` might not have reached an operation
+        // (think `UnknownOperationException`).
+        builder = builder.extension(crate::ResponseExtensions::new("TODO", "TODO"));
```

Comment:
What do we do here? Do we make it so that exceptions take in an _optional_ (namespace, operation name) pair so that we can fill this in? It must be optional because some exceptions could be raised before code execution gets to a particular operation deserialization routine (e.g. `UnknownOperationException`).

Reply:
What if we make it very simple and use "unknown", "unknown" in case we did not reach the point where we know operation and namespace? I am saying this because from a metrics perspective, you want to emit that an error happened, but you don't really know for which operation, hence unknown.

Reply:
Let's use this summary thread to talk about the extensions mess https://github.com/awslabs/smithy-rs/pull/1233/files#r818792743